### ISSUE TITLE
Enable Plugins in Modules

### ISSUE THREAD

**Issue by ghost** [vuejs/vuex#467]
Currently you cannot define a plugin when using a module store. I would like to be able to define plugins as part of a module that is being registered. I have built a `user` store that accepts the id, email and key for the user. I would like to define a plugin that saves the email to build remember me functionality.

```
store.registerModule('user', {
  state: {
    id: null,
    email: window.localStorage.getItem('email') || null,
    key: null
  },

  mutations: {
    register: (state, payload) => state.email = payload.email
  },

  plugins: [
    (store) => {
      store.subscribe((mutation, state) => {
        if (mutation.type === 'register') {
          window.localStorage.setItem('email', state.email)
        }
    }
  ]
})
```

**Comment by ktsn** [comment]
Sounds good to me 🙂 
One thing we may need to consider is how to unregister in-module plugins. As modules can be unregistered dynamically, in-module plugins can be also unregistered, I think.

**Comment by yanick** [comment]
For what it's worth, I have a first stab at that at https://github.com/yanick/vuex/tree/plugins-in-modules

Since a plugin can subscribe/watch to many things, I assume that plugins that want to be able to unregister as their module will be have to return an anonymous function that will call the unregistering function of all those subscribes/watches.

Of course, if the plugin only does a single subscribe, like
```
plugins: {
   foo: store => store.subscribe( ... )
}
```

it'll all magically work (as the return value of the subscribe is passed directly as the return value of the plugin.

**Comment by snaptopixel** [comment]
I'm not sure "plugins" are appropriate for this type of behavior (I actually found this issue trying to implement a similar strategy as @JBodkin).

It may be better to have lifecycle hooks instead:

```js
const myModule = {
  mutations: …
  actions: …
  registered(store) {
    getSomethingFromSomewhere().then(thing => store.commit('GOT_THING', thing));
    this.unsubscribe = store.subscribe(…);
  }
  unregistered(store) {
    this.unsubscribe();
  }
```

**Comment by snaptopixel** [comment]
FWIW, in my project I just did this in the store:
```js
plugins: [
  store => store.dispatch('STORE_INIT')
],
```
Then any modules can have a `STORE_INIT` action that can do whatever setup needed. In my case it was calling `firebase.auth().onAuthStateChanged` to check for existing auth.

**Comment by yanick** [comment]
> Then any modules can have a STORE_INIT action that can do whatever setup needed.

True. Of course, the actions don't pass the store itself as the context, but that can be worked around:

```
plugins: [
  store => store.dispatch('STORE_INIT', { store })
],
```

**Comment by snaptopixel** [comment]
OIC @yanick, so the "context" could be the module instead of the store (where the dispatch originated)? Still getting my feet wet with Vue/Vuex...

**Comment by yanick** [comment]
@snaptopixel  Kinda. The first argument passed to the action is still the context, which has the local state, the rootState, dispatch and commit, but if we want to do a subscribe or watch as part of the action, we need the store object, which in my snippet above is simply passed as the payload.

**Comment by snaptopixel** [comment]
Appreciate the "context" @yanick, thanks! I've updated my code based on your example and I've also renamed the action `$init` which feels a bit better :)

**Comment by yanick** [comment]
@snaptopixel  My pleasure. I'm also new to Vue, and it helps to compare notes with other peeps. :-)

**Comment by andyjessop** [comment]
Has there been any further work on this? I'm dying to be able to have plugins in modules for a current project. I want to be able to initialise each module with a bunch of data subscriptions, but I don't want to have to call each `init` action from the main store (the stores are namespaced), and even then, I don't want to have to explicitly set up each subscription within an `init` function.

So I want a plugin that can look in a local `subscriptions.js` file - cycle through, and make the subscriptions. But for each module, not the main store.

**Comment by bradstewart** [comment]
@andyjessop For what it's worth, I usually take the exact opposite approach: add a root-level plugin that dynamically registers it's associated module.

```js
export default function myPlugin () {
  // ... setup subscriptions, whatever else you need to initialize
  return (store) => {
    store.registerModule('myModule', {
      // ... module definition
    })
  }
}
```

If you have more specific timing requirements, need to wait on something else to happen before subscribing, etc. you can subscribe to mutations before running your initialization and registering the module. Should be able to avoid an `init` action per module since the plugin(s) is(are)n't namespaced.

**Comment by rp-21** [comment]
@snaptopixel  @yanick 

Just tried your Store.dispatch

hoping you figured this out already 

if you would like to trigger the actions and mutations of modules should use the entire namespace style mention

**Comment by shredmaster** [comment]
I need to do something similar and the helper function below does the trick for me.

```
// plugin.js
const email = (store:Store<any>, state: State, payload:any) => {
}
const password = (store:Store<any>, state: State, payload:any) => {
}
const handler = {email, password};
export const HelpPlugin: Plugin<State> = createModulePlugin("login", handler);

// helper.js
export function  createModulePlugin(module, handlers) {
    return (store: Store<any>) => {
        store.subscribe((mutation: any, state:any) => {
            let {type, payload} = mutation;
            let path : Array<string> = type.split('/');
            if(path.length !== 2) return;
            let moduleName = path.shift();
            if(moduleName !== module) return;
            let moduleState = state[moduleName];
            if(!moduleState) return;
            let handlerName = path.shift();
            let handler = handlers[handlerName];
            if(!handler) {
                return;
            }
            handler(store, moduleState, payload)
        })
    };
}
```

**Comment by zhDmitry** [comment]
if you need init method

`function registerInit(vueModule, prefix, acc) {
  Object.keys(vueModule.modules || {}).forEach((key) => {
    const nestedPrefix = prefix ? `${prefix}/${key}` : key;
    acc.push(nestedPrefix);
    registerInit(vueModule.modules[key], nestedPrefix, acc);
  });
}
function addLifehooks(vueModule, prefix) {
  const moduleInit = _.get(vueModule, 'actions.init', _.noop);
  const modulesWithInit = [];
  function init(ctx) {
    moduleInit(ctx);
    modulesWithInit.forEach((el) => {
      ctx.dispatch(`${el}/init`);
    });
  }
  registerInit(vueModule, prefix, modulesWithInit);
  _.set(vueModule, 'actions.init', init);
  return vueModule;
}
`

**Comment by zhDmitry** [comment]
```function registerInit(vueModule, prefix, acc) {
  Object.keys(vueModule.modules || {}).forEach((key) => {
    const nestedPrefix = prefix ? `${prefix}/${key}` : key;
    acc.push(nestedPrefix);
    registerInit(vueModule.modules[key], nestedPrefix, acc);
  });
}
function addLifehooks(vueModule, prefix) {
  const moduleInit = _.get(vueModule, 'actions.init', _.noop);
  const modulesWithInit = [];
  function init(ctx) {
    moduleInit(ctx);
    modulesWithInit.forEach((el) => {
      ctx.dispatch(`${el}/init`);
    });
  }
  registerInit(vueModule, prefix, modulesWithInit);
  _.set(vueModule, 'actions.init', init);
  return vueModule;
}
```

**Comment by frlinw** [comment]
Lifecycle hooks would be convenients, is there a plan to implement it or not at all ? (the tag seems to be outdated)

**Comment by Aferz** [comment]
Any update on this issue? Is there any plan to implement lifecycle hooks? That would be a really nice approach for `register/unregister` module `mutation/actions` subscriptions encapsulated in its own file.

**Comment by ksnyde** [comment]
@bradstewart liked your approach, do you have any examples you could point to in the public domain?

**Comment by cpmech** [comment]
A nice solution to this would be awesome...

**Comment by dylanbhughes** [comment]
What's the current status on this issue? Would the core team be open to contribution here? I'd love to help

**Comment by mkapiczy** [comment]
Any progress here? I would also love that feature. 

In my current application I am creating a module for every domain type object (e.g. user / application etc.). The server I am getting the data from enables websocket subscriptions per type, so I would like to have a websocketPlugin per module - opening appropriate socket subscription, and committing changes to the module's state.

**Comment by hackel** [comment]
This would be useful to use third-party vuex plugins with dynamically registered modules that only get loaded in one specific part of an app. (Especially for code splitting.)

**Comment by Frontrider** [comment]
A specific usecase:
There is a plugin that puts your state into local storage, but there's only one module I want to store. I can write my store that it uses the local storage, sure but I'd rather have that feature as a module.

**Comment by kiaking** [comment]
Closing due to the age.

It might be nice to have plugin only applied to the module, but I don't think this is the way to go. It's better for the plugin to provide such option (make selectable module to apply its feature).

**Comment by Frontrider** [comment]
I would say that it is a good thing, as it opens up the possibility to use some existing plugins, without having to rewrite them. It provides a very high level of flexibility.

Our use case did not enter production, but is still hanging in the air. We can try to find an another solution tough.

**Comment by egoson** [comment]
i use this aproach to handle mutation from requared module

// mutationTypes.js
```
export default {
  setUser: '[USER] Set User',
  removeUser: '[USER] Remove User'
}
```

// userModule.js
```
import mutationTypes form './mutationTypes'

export default {
  mutations: {
    [mutationTypes.setUser](state, payload) {
      state.user = payload
    }
  }
}
```

and in the plugin i check mutations by string

// mutationHandler.js
```
export default (module) => {
  return (store) => {
    store.subscribe((mutation, store) => {
      if (mutation.type.startsWith(module)) {
        **do somethings**
      } else {
      }
    })
  }
}
```

// index.js

```
...
plugins: [
    mutationHandler(['[USER]']),
  ],
```
